### PR TITLE
Register observer & correct clock setting

### PR DESCRIPTION
**Pull request by fjwillemsen**
This pull request adds a built-in Register Observer. This observer works for the PyCUDA, CuPy, and CUDA-Python backends. On unsupported backends, it gives a `NotImplementedError`. 
In addition, the pull request improves the efficiency with which clocks are set, and does not count the time spent doing so towards the benchmark time.

### CODE REVIEW

File: test/test_util_functions.py [hunk 1]
```diff
@@ -146,6 +146,19 @@ def test_get_thread_block_dimensions():
...
 
 
+def test_to_valid_nvrtc_gpu_arch_cc():
+    assert to_valid_nvrtc_gpu_arch_cc("89") == "89"
+    assert to_valid_nvrtc_gpu_arch_cc("88") == "87"
+    assert to_valid_nvrtc_gpu_arch_cc("86") == "80"
+    assert to_valid_nvrtc_gpu_arch_cc("40") == "52"
+    assert to_valid_nvrtc_gpu_arch_cc("90b") == "90a"
+    assert to_valid_nvrtc_gpu_arch_cc("91c") == "90a"
+    assert to_valid_nvrtc_gpu_arch_cc("10123001") == "52"
```
Comment (csbnw) [inline test/test_util_functions.py:156]: Order these in chronological order, e.g. put `52` as first, followed by `80` and so on. I wonder why Pascal, Volta and Turing are not tested here.

File: kernel_tuner/backends/cupy.py [hunk 2]
```diff
@@ -46,6 +47,7 @@ def __init__(self, device=0, iterations=7, compiler_options=None, observers=None
         self.devprops = dev.attributes
         self.cc = dev.compute_capability
         self.max_threads = self.devprops["MaxThreadsPerBlock"]
+        self.cache_size_L2 = self.devprops["L2CacheSize"]
```
Comment (csbnw) [inline kernel_tuner/backends/cupy.py:50]: After reading the PR text, it comes as a surprise to me that L2 related things have been changed.

File: kernel_tuner/backends/cupy.py [hunk 3]
```diff
@@ -124,6 +126,7 @@ def compile(self, kernel_instance):
         compiler_options = self.compiler_options
         if not any(["-std=" in opt for opt in self.compiler_options]):
             compiler_options = ["--std=c++11"] + self.compiler_options
+        # CuPy already sets the --gpu-architecture by itself, as per https://github.com/cupy/cupy/blob/20ccd63c0acc40969c851b1917dedeb032209e8b/cupy/cuda/compiler.py#L145
```
Comment (csbnw) [inline kernel_tuner/backends/cupy.py:129]: Is there a recommend version of CuPy for KernelTuner? If so, consider replacing the link to e.g. `https://github.com/cupy/cupy/blob/v13/cupy/cuda/compiler.py#L145` for `v13` of CuPy. Something like https://pypi.org/project/bump2version/ may be helpful to keep such version numbers up to date going forward.

File: kernel_tuner/backends/nvcuda.py [hunk 4]
```diff
@@ -192,6 +198,11 @@ def compile(self, kernel_instance):
             )
             cuda_error_check(err)
 
+            # get the number of registers per thread used in this kernel
+            num_regs = cuda.cuFuncGetAttribute(cuda.CUfunction_attribute.CU_FUNC_ATTRIBUTE_NUM_REGS, self.func)
+            assert num_regs[0] == 0, f"Retrieving number of registers per thread unsuccesful: code {num_regs[0]}"
+            self.num_regs = num_regs[1]
```
Comment (csbnw) [inline kernel_tuner/backends/nvcuda.py:204]: Again, I wonder why this is part of this MR. It is not about setting the correct clock.

File: kernel_tuner/backends/opencl.py [hunk 5]
```diff
@@ -45,6 +45,10 @@ def __init__(
         self.max_threads = self.ctx.devices[0].get_info(
             cl.device_info.MAX_WORK_GROUP_SIZE
         )
+        # TODO the L2 cache size request fails
+        # self.cache_size_L2 = self.ctx.devices[0].get_info(
+        #     cl.device_affinity_domain.L2_CACHE
+        # )
```
Comment (csbnw) [inline kernel_tuner/backends/opencl.py:51]: Open a new issue to keep track of this?

File: kernel_tuner/core.py [hunk 6]
```diff
@@ -340,14 +340,62 @@ def __init__(
         if not quiet:
             print("Using: " + self.dev.name)
 
-    def benchmark_default(self, func, gpu_args, threads, grid, result):
-        """Benchmark one kernel execution at a time"""
+        if lang.upper() not in ['OPENCL', 'C', 'FORTRAN']:
+            # flush the L2 cache, inspired by https://github.com/pytorch/FBGEMM/blob/eb3c304e6c213b81f2b2077813d3c6d16597aa97/fbgemm_gpu/bench/verify_fp16_stochastic_benchmark.cu#L130
```
Comment (csbnw) [inline kernel_tuner/core.py:344]: I can see why this may be useful. It also explains why you had to make some L2 related changes above. I would highly recommend moving these changes to a separate PR.

File: kernel_tuner/runners/sequential.py [hunk 7]
```diff
@@ -100,7 +100,7 @@ def run(self, parameter_space, tuning_options):
                 params = process_metrics(params, tuning_options.metrics)
 
             # get the framework time by estimating based on other times
-            total_time = 1000 * (perf_counter() - self.start_time) - warmup_time
+            total_time = 1000 * ((perf_counter() - self.start_time) - warmup_time)    # TODO is it valid that we deduct the warmup time here?
```
Comment (csbnw) [inline kernel_tuner/runners/sequential.py:103]: It depends on when `self.start_time` (before or after the warm-up?) is set and whatever `perf_counter()` returns. I don't know about these KernelTuner details to give a definitive answer, though.

File: kernel_tuner/util.py [hunk 8]
```diff
@@ -221,7 +221,7 @@ def check_block_size_names(block_size_names):
         if not isinstance(block_size_names, list):
             raise ValueError("block_size_names should be a list of strings!")
         if len(block_size_names) > 3:
-            raise ValueError("block_size_names should not contain more than 3 names!")
+            raise ValueError(f"block_size_names should not contain more than 3 names! ({block_size_names=})")
```
Comment (csbnw) [inline kernel_tuner/util.py:224]: Useful for debugging, but should it be included in this PR?

File: kernel_tuner/util.py [hunk 9]
```diff
@@ -570,6 +570,24 @@ def get_total_timings(results, env, overhead_time):
     return env
 
 
+def to_valid_nvrtc_gpu_arch_cc(compute_capability: str) -> str:
+    """Returns a valid Compute Capability for NVRTC `--gpu-architecture=`, as per https://docs.nvidia.com/cuda/nvrtc/index.html#group__options."""
+    valid_cc = ['50', '52', '53', '60', '61', '62', '70', '72', '75', '80', '87', '89', '90', '90a']    # must be in ascending order, when updating also update test_to_valid_nvrtc_gpu_arch_cc
```
Comment (csbnw) [inline kernel_tuner/util.py:575]: Here you do have the Pascal, Volta and Turing architectures! Can't you put these in some global list and use them in the tests as well to avoid having this list duplicated?

File: kernel_tuner/util.py [hunk 10]
```diff
@@ -570,6 +570,24 @@ def get_total_timings(results, env, overhead_time):
...
+    # if the compute capability does not match, scale down to the nearest matching
+    subset_cc = [cc for cc in valid_cc if compute_capability[0] == cc[0]]
+    if len(subset_cc) > 0:
+        # get the next-highest valid CC
+        highest_cc_index = max([i for i, cc in enumerate(subset_cc) if int(cc[1]) <= int(compute_capability[1])])
+        return subset_cc[highest_cc_index]
+    # if all else fails, return the default 52
+    return '52'
+
+
```
Comment (csbnw) [inline kernel_tuner/util.py:590]:
```suggestion
```

File: kernel_tuner/util.py [hunk 11]
```diff
@@ -570,6 +570,24 @@ def get_total_timings(results, env, overhead_time):
...
+    if len(compute_capability) < 2:
+        raise ValueError(f"Compute capability '{compute_capability}' must be at least of length 2, is {len(compute_capability)}")
+    if compute_capability in valid_cc:
+        return compute_capability
+    # if the compute capability does not match, scale down to the nearest matching
+    subset_cc = [cc for cc in valid_cc if compute_capability[0] == cc[0]]
+    if len(subset_cc) > 0:
+        # get the next-highest valid CC
+        highest_cc_index = max([i for i, cc in enumerate(subset_cc) if int(cc[1]) <= int(compute_capability[1])])
+        return subset_cc[highest_cc_index]
```
Comment (csbnw) [inline kernel_tuner/util.py:586]: This seems like a rather complicated way of trying to match the input `compute_capability` to something in `valid_cc`. I am sure there must be a way to make it simpler and more readable.

**Review by csbnw — COMMENTED** [review]
I happened to stumble upon this PR and out of curiosity had a look at the changes and couldn't help to add some comments. My main 'complaint't is that some scope creep seems to have occurred regarding the L2 flushing. In my opinion, it should really be moved into a separate PR.

File: test/test_util_functions.py (hunk 1 above)
Comment (fjwillemsen) [reply test/test_util_functions.py:156]: The idea was to just test the edge cases for the mapping, not specific architectures or all architectures exhaustively.

File: kernel_tuner/backends/cupy.py (hunk 3 above)
Comment (fjwillemsen) [reply kernel_tuner/backends/cupy.py:129]: Good point, it should indeed be version independent. I'll change it to `main` instead for now.

File: kernel_tuner/util.py (hunk 10 above)
Comment (fjwillemsen) [reply kernel_tuner/util.py:590]: All functions in `util.py` are separated by two lines

File: kernel_tuner/util.py (hunk 11 above)
Comment (fjwillemsen) [reply kernel_tuner/util.py:586]: I agree, the problem is we can't use integer comparison because it's also possible to have e.g. `90a`, and we can only match within the major number. Any suggestions?

Comment (csbnw) [reply kernel_tuner/util.py:586]: How about using difflib, e.g. the [`get_close_matches` ](https://docs.python.org/3/library/difflib.html#difflib.get_close_matches?) function.
